src/Symfony/Component/TypeInfo/Type/UnionType.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\TypeInfo\Type;
+
+use Symfony\Component\TypeInfo\Type;
+use Symfony\Component\TypeInfo\TypeIdentifier;
+
+/**
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ * @author Baptiste Leduc <baptiste.leduc@gmail.com>
+ *
+ * @template T of Type
+ */
+final class UnionType extends Type
+{
+    /**
+     * @use CompositeTypeTrait<T>
+     */
+    use CompositeTypeTrait;
+
+    public function asNonNullable(): Type
+    {
+        $types = array_values(array_filter($this->getTypes(), fn (Type $t): bool => !$t->isA(TypeIdentifier::NULL)));
```

Comment:
Indeed, `mixed` is a special type... As stated in the documentation, it says that "It is equivalent to the union type object|resource|array|string|float|int|bool|null".

Therefore, instead of having a `mixed` built-in type, the static factory could convert it to the proper union type, so that there is no ambiguity anymore. WDYT?

Reply:
By doing that, it gives the following:
```php
$type = Type::union(Type::int(), Type::mixed());
(string) $type; // array|bool|float|int|null|object|resource|string
$type->isNullable(); // true

$type = $type->asNonNullable();
(string) $type; // array|bool|float|int|object|resource|string
$type->isNullable(); // false
```

Reply:
It's an interesting idea, but I don't know if it will work well as a project-wide solution. I can't say how the component is going to be used by developers, but perhaps there are use cases where we would want to know exactly whether the type is written as `mixed` or as union. Without the `mixed` built-in type, we will lose this information if I understand your idea correctly. 

But as for that narrow scope of the `asNonNullable` method, it seems to me that this conversion and then stripping `null` would make sense. Again, I can't tell right now how this method is going to be used so I'm curious what others think about this, perhaps their perspective is different.

Reply:
I can see one unexpected behavior with that approach: `(string) $type` will give `array|bool|float|int|null|object|resource|string` instead of the expected `mixed`.

Therefore, I introduced again the `mixed` built-in type and convert it to its related union type only when really needed, which is when `asNonNullable` is involved.

Reply:
don't know if it's easier but you can also have it expanded to `array|bool|float|int|object|resource|string` inside, and transforming that to `mixed` when printing its name if it's an union matching that, or `mixed|A|B|C` if there's more

to me it feels like it would make the internal logic easier to deal with in the future if it doesn't have to care about that special case